pennylane/devices/qubit/apply_operation.py

```diff
@@ -0,0 +1,179 @@
+        state (array[complex]): input state
+
+    Returns:
+        array[complex]: output_state
+    """
+    mat = op.matrix()
+
+    total_indices = len(state.shape)
+    num_indices = len(op.wires)
+
+    state_indices = ABC[:total_indices]
+    affected_indices = "".join(ABC[i] for i in op.wires)
+
+    new_indices = ABC[total_indices : total_indices + num_indices]
+
+    new_state_indices = state_indices
+    for old, new in zip(affected_indices, new_indices):
+        new_state_indices = new_state_indices.replace(old, new)
+
+    einsum_indices = f"{new_indices}{affected_indices},{state_indices}->{new_state_indices}"
+
+    reshaped_mat = math.reshape(mat, [2] * (num_indices * 2))
+
+    return math.einsum(einsum_indices, reshaped_mat, state)
+
+
+def apply_operation_tensordot(op: qml.operation.Operator, state):
+    """Apply ``Operator`` to ``state`` using ``math.tensordot``. This is more efficent at higher qubit
+    numbers.
+
+    Args:
+        op (Operator): Operator to apply to the
+        state (array[complex]): input state
+
+    Returns:
+        array[complex]: output_state
+    """
+    mat = op.matrix()
+    total_indices = len(state.shape)
+    num_indices = len(op.wires)
+    reshaped_mat = math.reshape(mat, [2] * (num_indices * 2))
+    axes = (tuple(range(num_indices, 2 * num_indices)), tuple(op.wires))
+
+    tdot = math.tensordot(reshaped_mat, state, axes=axes)
+
+    unused_idxs = [i for i in range(total_indices) if i not in op.wires]
+    perm = list(op.wires) + unused_idxs
+    inv_perm = math.argsort(perm)
+
+    return math.transpose(tdot, inv_perm)
+
+
+@singledispatch
+def apply_operation(op: qml.operation.Operator, state):
+    """Apply and operator to a given state.
+
+    Args:
+        op (Operator): The operation to apply to ``state``
+        state (ndarray): The starting state.
+
+    Returns:
+        ndarray: output state
+
+    .. warning::
+
+        ``apply_operation`` is an internal function, and thus subject to change without a deprecation cycle.
+
+    .. warning::
+        ``apply_operation`` applies no validation to its inputs.
+
+        This function assumes that the wires of the operator correspond to indices
+        of the state. See :func:`~.map_wires` to convert operations to integer wire labels.
+
+        The shape of state should be ``[2]*num_wires``.
+
+    This is a ``functools.singledispatch`` function, so additional specialized kernels
+    for specific operations can be registered like:
+
+    .. code-block:: python
+
+        @apply_operation.register
+        def _(op: type_op, state):
+            # custom op application method here
+
+    **Example:**
+
+    >>> state = np.zeros((2,2))
+    >>> state[0][0] = 1
+    >>> state
+    tensor([[1., 0.],
+        [0., 0.]], requires_grad=True)
+    >>> apply_operation(qml.PauliX(0), state)
+    tensor([[0., 0.],
+        [1., 0.]], requires_grad=True)
+
+    """
+    if len(op.wires) < 3:
```

Comment:
From another PR from @dwierichs, einsum is faster until 6 qubits for Jax (default mixed). It would be great if we can benchmark that for the new default and also condition on the interface if it is also the case.

Reply:
+1 on this number, forgot to ask. also, it's generally good to define "magic numbers" as constants so we know what they mean. something like `EINSUM_WIRECOUNT_PERF_THRESHOLD = 3`

Reply:
> until 6 qubits for Jax (default mixed)

Indeed I think it might be nice to benchmark this _a little_ (will put up the benchmark in the [default mixed case](https://github.com/PennyLaneAI/pennylane/pull/3584) later today)

Reply:
bechmarking info attached in the introduction to this PR.